protostar/commands/test/test_result_formatter.py

```diff
@@ -0,0 +1,248 @@
+from pathlib import Path
+from typing import Callable, Dict, List
+
+from protostar.commands.test.test_environment_exceptions import ExceptionMetadata
+from protostar.commands.test.test_output_recorder import OutputName, format_output_name
+from protostar.protostar_exception import UNEXPECTED_PROTOSTAR_ERROR_MSG
+from protostar.utils.log_color_provider import LogColorProvider
+
+from .test_results import (
+    BrokenTestSuiteResult,
+    FailedFuzzTestCaseResult,
+    FailedTestCaseResult,
+    PassedFuzzTestCaseResult,
+    PassedTestCaseResult,
+    TestResult,
+    UnexpectedBrokenTestSuiteResult,
+)
+
+LogCallback = Callable[[str], None]
+
+
+class TestResultFormatter:
```

Comment:
discussed this class privately in office

Reply:
After giving it a thought, I noticed the following issues with your suggestion:
- presenters would have to be a class with a single method — this is what we want to avoid
- how should presenters be kept in files?
  - all in one? — the file grows quickly when we decide to add more methods to a presenter 
  - each presenter in a separate file? — lack of consistency with the `test_result.py`
    - splitting `test_result.py` on multiple files and keeping model and presenter in one folder? — cohesive part of a domain is split and harder to read because of unnecessary coupling with the details

Regarding the `if isinstance()` chain. It can be considered a code smell suggesting using polymorphism when it's used on objects.
 
> Objects hide their data behind abstractions and expose functions that operate on that data. Data structures expose their data and have no meaningful functions.

> Procedural code (code using data structures) makes it easy to add new functions without changing the existing data structures. Object oriented code, on the other hand, makes it easy to add new classes without chaging existing functions.

`TestResults` is the Protostar domain, and it should be protected (the dependency rule). `TestResults` are unlikely to grow further. Even if they do, code shouldn't break due to the hierarchical nature of `TestResults`.

Reply:
> * presenters would have to be a class with a single method — this is what we want to avoid

So we could use plain functions here, IMHO. At least until we introduce multiple output formats.

> how should presenters be kept in files?
> * all in one? — the file grows quickly when we decide to add more methods to a presenter
> * each presenter in a separate file? — lack of consistency with the `test_result.py`
>   
>   * splitting `test_result.py` on multiple files and keeping model and presenter in one folder? — cohesive part of a domain is split and harder to read because of unnecessary coupling with the details

I would go with each presenter in a separate file because they are already big. I don't see a big problem with a lack of consistency with `test_result.py`. It is, but I think this tradeoff is surely acceptable.

> Regarding the `if isinstance()` chain. It can be considered a code smell suggesting using polymorphism when it's used on objects.
> 
> > Objects hide their data behind abstractions and expose functions that operate on that data. Data structures expose their data and have no meaningful functions.
> 
> > Procedural code (code using data structures) makes it easy to add new functions without changing the existing data structures. Object oriented code, on the other hand, makes it easy to add new classes without chaging existing functions.
> 
> `TestResults` is the Protostar domain, and it should be protected (the dependency rule). `TestResults` are unlikely to grow further. Even if they do, code shouldn't break due to the hierarchical nature of `TestResults`.

I need to think about this more. This doesn't look so much smelly to me, but maybe you are right. Although I would definitely still raise instead of returning `""`.

Reply:
> So we could use plain functions here, IMHO. At least until we introduce multiple output formats.

Such code won't be easy to extend (Open-Closed principle violation) because every caller of that method would have to be updated. Devs will be more willing to add a new formatting methods analogously, which leads to mess in a `test_result.py` once again.

> Although I would definitely still raise instead of returning "".

I think about creating a `test_result` formatter (e.g. `[UNEXPECTED_TEST_RESULT] filepath`) add using `assert` instead of exception. (I'm not sure if Protostar should crash on production.)

Reply:
> I think about creating a `test_result` formatter (e.g. `[UNEXPECTED_TEST_RESULT] filepath`) add using `assert` instead of exception. (I'm not sure if Protostar should crash on production.)

It definitely should. It's a programmer bug to forget to include this and we do not have any compile-time means of verifying this.

> Such code won't be easy to extend (Open-Closed principle violation) because every caller of that method would have to be updated. Devs will be more willing to add a new formatting methods analogously, which leads to mess in a `test_result.py` once again.

I don't understand this sentence. What would be the incentive for extending presenter classes instead of composing lower level functions? Even more, I don't see how for example these presenters could extend each other, as they vastly differ:
- Null presenter
- CLI text presenter
- JSON presenter
- TeamCity messages presenter

Reply:
> I don't understand this sentence. What would be the incentive for extending presenter classes instead of composing lower level functions

NVM, I thought for some reason you wanted to use a method of the Test Result as a temporary solution rather than a function. I'll convert formatters to functions in the next PR.

Reply:
Converted to functions in this PR. Refinement will be in another.